net_handler/tcp_conn.c

```diff
@@ -254,7 +254,25 @@ static int recv_new_msg (int conn_fd, int challenge_fd)
 			log_printf(ERROR, "Cannot unpack dev_data msg");
 			return -2;
 		}
-		log_printf(WARN, "Received device data msg which has no implementation");
+		for (int i = 0; i < dev_data_msg->n_devices; i++) {
+			Device* req_device = dev_data_msg->devices[i];
+			device_t* act_device = get_device(req_device->type);
+			if (act_device == NULL) {
+				log_printf(ERROR, "Invalid device subscription: device type %d is invalid", req_device->type);
+				continue;
+			}
+			uint32_t requests = 0;
+			for (int j = 0; j < req_device->n_params; j++) {
+				if (req_device->params[i]->val_case == PARAM__VAL_IVAL) {
```

Comment:
I think I told Dawn to use the boolean to request subscriptions (so if student requests param `i` then the value of param `i` coming in will be `True`, and vice versa for `False`). You can assume that any nonexistent params or params that aren't specified in the message are `False`.

Reply:
Ok changed it to check the boolean value instead.